### ISSUE TITLE
Major regression: scan fails if a path contains weird characters

### ISSUE THREAD

**Issue by pombredanne** [nexB/scancode-toolkit#390]
This seems to be a side effects or results caching
```
>>> a='Izgradnja sufiksnog polja kori\xc5\xa1tenjem K\xc3\xa4rkk\xc3\xa4inen \xe2\x80\x93 Sandersovog algoritma.pdf'
>>> with open(a, 'wb') as o:
...  o.write('\n')
... 
```
scancode -i Izgradnja\ sufiksnog\ polja\ korištenjem\ Kärkkäinen\ –\ Sandersovog\ algoritma.pdf 
Scanning files for: infos with 1 process(es)...
Scanning files...
  [####################################]  1                                                                                            
Scanning done.
Scan statistics: 1 files scanned in 0s.
Scan options:    infos with 1 process(es).
Scanning speed:  8.48 files per sec.
Scanning time:   0s.
Indexing time:   0s.
Saving results.
{
  "scancode_notice": "Generated with ScanCode and provided on an \"AS IS\" BASIS, WITHOUT WARRANTIES\nOR CONDITIONS OF ANY KIND, either express or implied. No content created from\nScanCode should be considered or used as legal advice. Consult an Attorney\nfor any legal advice.\nScanCode is a free software code scanning tool from nexB Inc. and others.\nVisit https://github.com/nexB/scancode-toolkit/ for support and download.",
  "scancode_version": "2.0.0rc2",
  "files_count": 1,
  "files": Traceback (most recent call last):
  File ".../bin/scancode", line 9, in <module>
    load_entry_point('scancode-toolkit', 'console_scripts', 'scancode')()
  File ".../local/lib/python2.7/site-packages/click/core.py", line 664, in __call__
    return self.main(*args, **kwargs)
  File ".../src/scancode/utils.py", line 62, in main
    standalone_mode=standalone_mode, **extra)
  File ".../local/lib/python2.7/site-packages/click/core.py", line 644, in main
    rv = self.invoke(ctx)
  File ".../local/lib/python2.7/site-packages/click/core.py", line 837, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File ".../local/lib/python2.7/site-packages/click/core.py", line 464, in invoke
    return callback(*args, **kwargs)
  File ".../src/scancode/cli.py", line 312, in scancode
    save_results(files_count, results, format, input, output_file)
  File ".../src/scancode/cli.py", line 617, in save_results
    json.dump(meta, output_file, indent=2 * ' ', iterable_as_array=True)
  File ".../local/lib/python2.7/site-packages/simplejson/__init__.py", line 276, in dump
    for chunk in iterable:
  File ".../local/lib/python2.7/site-packages/simplejson/encoder.py", line 677, in _iterencode
    for chunk in _iterencode_dict(o, _current_indent_level):
  File ".../local/lib/python2.7/site-packages/simplejson/encoder.py", line 634, in _iterencode_dict
    for chunk in chunks:
  File ".../local/lib/python2.7/site-packages/simplejson/encoder.py", line 689, in _iterencode
    for chunk in _iterencode_list(o, _current_indent_level):
  File ".../local/lib/python2.7/site-packages/simplejson/encoder.py", line 475, in _iterencode_list
    for value in lst:
  File ".../src/scancode/cache.py", line 254, in iterate
    file_info = self.get_info(path)
  File ".../src/scancode/cache.py", line 209, in get_info
    info_path = self.get_cached_info_path(path)
  File ".../src/scancode/cache.py", line 186, in get_cached_info_path
    keys = info_keys(path)
  File ".../src/scancode/cache.py", line 105, in info_keys
    return keys_from_hash(sha1(path).hexdigest())
UnicodeEncodeError: 'ascii' codec can't encode character u'\u0161' in position 30: ordinal not in range(128)

```

**Comment by pombredanne** [comment]
This can also happen earlier on with directories such as this Python string (which is new folder in russian)
```
'\xd0\x9d\xd0\xbe\xd0\xb2\xd0\xb0\xd1\x8f \xd0\xbf\xd0\xb0\xd0\xbf\xd0\xba\xd0\xb0'
```

**Comment by pombredanne** [comment]
extractcode should be checked too.

**Comment by pombredanne** [comment]
the build is failing but the main issue is minor: this is only a matter of scanned file date not being stable: let me fix this by wrapping the unicode test files in an archive. 
Another thing is that paths with weird unicode characters will not endup as the same path on all OSes. This could eventually be fixed but is a difficult issue as the ways file systems encode file names on each OS is markedly different: binary on Linux, some kind of unicode on Mac and yet another flavor of unicode on Windows.... plus the eventual locale issues. 

For reference, the name that is problematic is something I created on Windows in python this way:
```
with open(unichr(1000)+unichr(8000)+unichr(9000), 'wb') as of:
    of.write('some text')
```

**Comment by pombredanne** [comment]
This is working OK now in develop. There are new features and tests for 7zip archives that were added.

**Comment by sschuberth** [comment]
@pombredanne I still see this happening (with using ScanCode from commit 1e3ade025f9ef12db38c11ee849a4c44c8a68d61), though in a different code path:

    Traceback (most recent call last):
      File "/var/lib/jenkins/jobs/scancode/workspace/tools/scancode/bin/scancode", line 9, in <module>
        load_entry_point('scancode-toolkit', 'console_scripts', 'scancode')()
      File "/var/lib/jenkins/jobs/scancode/workspace/tools/scancode/local/lib/python2.7/site-packages/click/core.py", line 664, in __call__
        return self.main(*args, **kwargs)
      File "/var/lib/jenkins/jobs/scancode/workspace/tools/scancode/src/scancode/utils.py", line 64, in main
        standalone_mode=standalone_mode, **extra)
      File "/var/lib/jenkins/jobs/scancode/workspace/tools/scancode/local/lib/python2.7/site-packages/click/core.py", line 644, in main
        rv = self.invoke(ctx)
      File "/var/lib/jenkins/jobs/scancode/workspace/tools/scancode/local/lib/python2.7/site-packages/click/core.py", line 837, in invoke
        return ctx.invoke(self.callback, **ctx.params)
      File "/var/lib/jenkins/jobs/scancode/workspace/tools/scancode/local/lib/python2.7/site-packages/click/core.py", line 464, in invoke
        return callback(*args, **kwargs)
      File "/var/lib/jenkins/jobs/scancode/workspace/tools/scancode/src/scancode/cli.py", line 311, in scancode
        save_results(files_count, results, format, input, output_file)
      File "/var/lib/jenkins/jobs/scancode/workspace/tools/scancode/src/scancode/cli.py", line 598, in save_results
        output_file.write(as_template(scanned_files))
    UnicodeEncodeError: 'ascii' codec can't encode character u'\xef' in position 7352437: ordinal not in range(128)

So, while you seem to have fixed the issue for *reading* from Unicode paths, my issues seems to be about *writing* to Unicode paths, if I read the stacktrace correctly.

**Comment by pombredanne** [comment]
@sschuberth let me look at that right away. This may be the same bug that @ened experienced once but that I could not reproduce in #132 
In terms of what was printed out, this was happening at the end once the scan was complete and the output file was being created using either an HTML or HTML-APP output from the what the stacktrace says.

**Comment by sschuberth** [comment]
@pombredanne Thanks for jumping at it. FYI, in my case the output format was HTML.

**Comment by pombredanne** [comment]
@sschuberth I am sorry you are finding all these bugs but I am grateful that you find them too at the same time!
You do not happen to know which file could be the culprit by chance?

**Comment by sschuberth** [comment]
> You do not happen to know which file could be the culprit by chance?

I'm not certain, but I'll re-run with `-n 1` and `--verbose` to find out. Mean while, I found this in the output:

    Some files failed to scan properly. See scan for details:
      sources/cxx-stl/llvm-libc%2B%2B/libcxx/test/localization/locale.categories/category.numeric/locale.nm.put/facet.num.put.members/put_long_double.pass.cpp
      sources/cxx-stl/llvm-libc%2B%2B/libcxx/test/localization/locale.categories/category.numeric/locale.nm.put/facet.num.put.members/put_double.pass.cpp
      sources/cxx-stl/llvm-libc%2B%2Babi/libcxxabi/test/test_demangle.pass.cpp

I'm not sure it's related, but i wonder why `+` is escaped as `%2B` here...

**Comment by pombredanne** [comment]
Thanks! and your `scancode` CLI options were what?  FWIW, the + being URL encoded is part of the fix I applied for this bug here, e.g. URL-encoding path segments .... but this is too aggressive as this is a valid file name character on all OSes.

**Comment by sschuberth** [comment]
This is how I ran ScanCode:

    $ scancode --timeout 180 -n 2 -f html . /var/lib/jenkins/jobs/scancode/workspace/results/android.googlesource.com//platform/ndk/2016-03-30_932961a/ndk-scancode-1e3ade0.html

**Comment by pombredanne** [comment]
@sschuberth Thanks. So I guess the code you tried to scan was that: https://android.googlesource.com/platform/ndk/+/932961a and I can use that to reproduce.
I guess you ran the scan on Linux too.

**Comment by sschuberth** [comment]
Yes, both of your assumptions are correct.

**Comment by pombredanne** [comment]
And bingo I have reproduce it!

```
mkdir -p tmp/ndk
cd tmp/ndk
wget https://android.googlesource.com/platform/ndk/+archive/932961a.tar.gz
mkdir -p android.googlesource.com/platform/ndk-932961a.tar.gz
tar -xf 932961a.tar.gz -C android.googlesource.com/platform/ndk-932961a.tar.gz/
cd  ../..
. bin/activate

./scancode -lp -n6 -f html tmp/ndk/android.googlesource.com/platform/ndk-932961a.tar.gz/ ndk-copy.html
Scanning files for: licenses, packages with 6 process(es)...
Building license detection index...Done.
Scanning files...
  [####################################]  10324                                                                         
Scanning done.
Some files failed to scan properly. See scan for details:
 sources/cxx-stl/llvm-libc%2B%2Babi/libcxxabi/test/test_demangle.pass.cpp
 sources/cxx-stl/llvm-libc%2B%2B/libcxx/test/localization/locale.categories/category.numeric/locale.nm.put/facet.num.put.members/put_double.pass.cpp
Scan statistics: 10324 files scanned in 520s.
Scan options:    licenses, packages with 6 process(es).
Scanning speed:  20.92 files per sec.
Scanning time:   493s.
Indexing time:   27s.
Saving results.
Traceback (most recent call last):
  File "/home/pombreda/w421/scancode-toolkit/bin/scancode", line 9, in <module>
    load_entry_point('scancode-toolkit', 'console_scripts', 'scancode')()
  File "/home/pombreda/w421/scancode-toolkit/local/lib/python2.7/site-packages/click/core.py", line 664, in __call__
    return self.main(*args, **kwargs)
  File "/home/pombreda/w421/scancode-toolkit/src/scancode/utils.py", line 64, in main
    standalone_mode=standalone_mode, **extra)
  File "/home/pombreda/w421/scancode-toolkit/local/lib/python2.7/site-packages/click/core.py", line 644, in main
    rv = self.invoke(ctx)
  File "/home/pombreda/w421/scancode-toolkit/local/lib/python2.7/site-packages/click/core.py", line 837, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/pombreda/w421/scancode-toolkit/local/lib/python2.7/site-packages/click/core.py", line 464, in invoke
    return callback(*args, **kwargs)
  File "/home/pombreda/w421/scancode-toolkit/src/scancode/cli.py", line 311, in scancode
    save_results(files_count, results, format, input, output_file)
  File "/home/pombreda/w421/scancode-toolkit/src/scancode/cli.py", line 598, in save_results
    output_file.write(as_template(scanned_files))
UnicodeEncodeError: 'ascii' codec can't encode character u'\xef' in position 6650293: ordinal not in range(128)
```

**Comment by pombredanne** [comment]
With this I can now debug this nicely now.

**Comment by pombredanne** [comment]
And I have a simple quick way to reproduce now. This is because some license metadata contain Unicode as they can alright. But because the output file is not opened to write as UTF-8 by click but only instructed to write as binary, it tries to convert perfectly correct unicode to ascii with no encoding is specified and fails badly there. I will have a fix soon and I will also improve the error reporting to get a better stacktrace. 
```
./scancode -l  -f html src/licensedcode/data/licenses/bsd-unchanged.LICENSE bug.html
```

**Comment by pombredanne** [comment]
@sschuberth the #407 PR should fix the problem.
The key was adding an encoding to the `click.File` here: https://github.com/nexB/scancode-toolkit/pull/407/files#diff-1e37ecfb9a8c3fccd8ecdfb2ce51f591L261

**Comment by sschuberth** [comment]
Thanks a lot! I'll give it a try ASAP.

**Comment by pombredanne** [comment]
This is now fixed for good with #407
